src/renderer_gl.h

```diff
@@ -6,27 +6,41 @@
 #ifndef BGFX_RENDERER_GL_H_HEADER_GUARD
 #define BGFX_RENDERER_GL_H_HEADER_GUARD
 
-#define BGFX_USE_EGL (BGFX_CONFIG_RENDERER_OPENGLES && (0 \
-	|| BX_PLATFORM_ANDROID                                \
-	|| BX_PLATFORM_BSD                                    \
-	|| BX_PLATFORM_LINUX                                  \
-	|| BX_PLATFORM_NX                                     \
-	|| BX_PLATFORM_RPI                                    \
-	|| BX_PLATFORM_WINDOWS                                \
-	) )
+#ifndef BGFX_USE_EGL
```

Comment:
Why is this wrapped into ifndef? User should not be providing this by compiler options.

Reply:
Previously, the logic was to use EGL when using GLES and use the native backend (such as GLX, WGL, etc.) when using GL. Now, allowing EGL for GL as well introduces a choice. On platforms that support both the native backend and EGL, there needs to be a way to configure which one to use at compile time.

Alternatively, we could consider using EGL consistently across all platforms. However, we might not be ready to replace the other backends systematically, so having the ability to choose between the original native backend and EGL
(i compile with BGFX_USE_EGL=1 BGFX_USE_GLX=0)

Reply:
This change makes it too complicated... I think EGL should stay GLES only.

Can you focus your PR only on: "allow headless context creation"?

Reply:
Hello, I'm working with @issam3105 on this.

As said, our context is CI and headless linux servers, the main goal is to get away from GLX (can't be headless). Linux GLES support seems good, so GL is indeed not actually very important.
We'll get back to the drawing board on the PR to just focus on the headless part.

Reply:
> (*headless meaning no display server; might have a GPU still, but sometimes not even, using mesa sw rendering)

Yeah. Headless is useful.

Reply:
OpenGL over EGL is also useful if can be used everywhere. For example if you can prove that we would be able to drop GLX on most popular Linux distros, I would be interested to use EGL instead of GLX.

Reply:
I seem to be working on the same thing, only coming at it from trying to get MAME working with BGFX+Wayland

https://github.com/bkaradzic/bgfx/issues/2416#issuecomment-1595740516

Yes, please(!) deprecate or drop GLX support.  X11 EGL support works perfectly well for both GLES and OpenGL.  This has been the case for many years now.

Reply:
My approach is to make GLX support default to disabled at build time behind a --with-glx switch, and have --with-wayland require *not* --with-glx.  If GLX is disabled have X11 only support EGL for both GLES and OpenGL.

Reply:
My patch now builds for Wayland, including Vulkan support, but crashes in the same way as MAME.  So progress...! 

The Vulkan surface creation is currently #ifdef'd per platform, and it's impossible as it stands to have both X11 and Wayland in a single build.  I've added the VKWayland code, but even to support two separate builds BX_PLATFORM_LINUX needs to be split into maybe BX_PLATFORM_LINUX_X11 and BX_PLATFORM_LINUX_WAYLAND or there needs to be a define BGFX_USE_WAYLAND/X11 and have it build one variant or the other.  For now I've just used  "#elif 0" for the X11 code.

As far as I can tell, it's only the Vulkan surface creation support which needs build time selection between X11/Wayland.

Reply:
@sjnewbury 

> Yes, please(!) deprecate or drop GLX support. X11 EGL support works perfectly well for both GLES and OpenGL. This has been the case for many years now.

This is actually completely inaccurate. Maybe it's true if you consider only Linux distribution you're using. But I don't have Wayland on Linux Mint / Cinnamon (and I don't intend to switch from it).

Reply:
@bkaradzic I said "**X11** EGL support".  If it's not working on your distribution that's a pretty serious deficiency,  it's been supported in Mesa for OpenGL as well as GLES  for well over a decade.  It is of course a requirement for Wayland, but it there is no good reason to use GLX for X11 any more.

Reply:
I'll create a new PR with what I've been working on. Feel free to shoot it down.

Reply:
@sjnewbury EGL instead of GLX is fine. Just focus your PR to have minimal changes to achieve that.

Reply:
@bkaradzic Okay, I'll create a PR with just the EGL by default changes and separate the work to get Wayland working to a different PR.

Reply:
Once EGL is used, bgfx doesn't have to care what's behind it: thanks to EGL external platform system, one can use either display server specific surfaces (x11/wayland window handles) or "unmapped" GBM windows (as the name does not imply, those are offscreen).
Unfortunately, nvidia's GBM support is still sketchy, one must use pbuffers on the "device platform" instead. I'd like to come up with a better way to pass that information to bgfx. This patch assumes that if the window handle is NULL, a pbuffer is created, but there is no interface to define it's size (I believe it'll default to 0x0).